src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIChatClientMetadata.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.AI;
+
+internal sealed class OpenAIChatClientMetadata(string provider, Uri endpoint, string? modelId)
+    : ChatClientMetadata(provider, endpoint, modelId)
+{
+    private const int MaxMetadataCacheEntries = 1000;
+    private static readonly ConcurrentDictionary<string, ChatModelMetadata> _metadataByModelId = new();
+
+    // See https://platform.openai.com/docs/guides/structured-outputs
+    private static readonly string[] _supportedStructuredOutputModelIdPrefixes =
+    [
+        "gpt-4.5",
+        "o3-mini",
+        "o1",
+        "gpt-4o",
+        "gpt-4o-mini",
+    ];
+
+    // These are model IDs that would otherwise match one of the prefixes above,
+    // but need to be special-cased because they don't support structured output.
+    private static readonly string[] _unsupportedStructuredOutputModelIdPrefixes =
+    [
+        "o1-mini",
+    ];
+
+    public override Task<ChatModelMetadata> GetModelMetadataAsync(string? modelId = null, CancellationToken cancellationToken = default)
+    {
+        if (string.IsNullOrEmpty(modelId))
+        {
+            modelId = DefaultModelId;
+        }
+
+        // There could be a brief race here and we could add more than the intended max items to the cache.
+        // That wouldn't be a problem since it's an arbitrary limit to mitigate memory leaks if the application
+        // is doing something very strange (generating distinct model IDs per call).
+        var result = _metadataByModelId.Count > MaxMetadataCacheEntries
+            ? (_metadataByModelId.TryGetValue(modelId ?? string.Empty, out var entry) ? entry : CreateModelMetadata(modelId))
+            : _metadataByModelId.GetOrAdd(modelId ?? string.Empty, CreateModelMetadata);
```

Comment:
Count on a ConcurrentDictionary is unfortunately not as cheap as would be ideal. It'd be better to only check it once we know we've missed the cache, e.g.
```C#
if (!_metadataByModelId.TryGetValue(modelId, out var result)
{
    result = CreateModelMetadata(modelId);
    if (_metadataByModelId.Count < MaxMetadataCacheEntries)
    {
        result = _metadataByModelId.GetOrAdd(modelId, result);
    }
}

return result;
```

Also, I'm unclear what all these string.Empty's are for. If the modelId is null, we still want to use the cache?

Reply:
> Also, I'm unclear what all these string.Empty's are for. If the modelId is null, we still want to use the cache?

I'm not sure the modelId can actually be null. `OpenAIChatClient` throws if you pass a null/empty modelId constructor param, but there's a second constructor that gets modelId using private reflection, so I was unusure if there's some case where a `null` value has a meaning like "use a global default". Based on your comment I've changed this to skip the cache and return empty metadata if the resolve modelId is null/empty.

> Count on a ConcurrentDictionary is unfortunately not as cheap as would be ideal

Thanks. Updated.